### PR TITLE
AEROGEAR-9184

### PR DESCRIPTION
**Pull request by laurafitzgerald**
**JIRA**
https://issues.jboss.org/browse/AEROGEAR-9184

**Tasks**
- [x] Add some dependencies in order to ad the new api for the a MobileSecurityApp
- [x] Generate the stuff
- [x] Have the operator respond to the creation of the app cr, aka create the required SDK config
- [x] Put the right stuff in the Confg Map
- [x] Edit - Have the config map be created if the cr is in a namespace which is provided by the APP_NAMESPACE variable.

**Verification Steps**

`make cluster/clean`
`make cluster/prepare`
`make install`

Run the following to verify the _correct_ namespace:
`make example-app/apply`
Verify that the sdk config map is created in the `app-metrics-apps` namespace with the name `<cr-app-name>-metrics`
Verify the contents of the config map is `url` : `the route for the metrics service`

Run the following to verify the _incorrect_ namespace:
`make example-app/delete`
Change the APP_NAMESPACES variable in the Makefile to `app-metrics`
`make example-app/apply`

Verify the cr is created in the `app-metrics` ns
Verify that the sdk config map is **not** created in the `app-metrics` namespace
Verify that the operator produces a log with the following content `"app-metrics/example-app","error":"The app cr example-app was created in a namespace which is not present in the APP_NAMESPACES environment variable provided to the operator","`

### CODE REVIEW

File: deploy/operator.yaml
Comment (laurafitzgerald) [inline deploy/operator.yaml]: This is pushed for convenience. The dev image is built and pushed for testing. I'll remove it before merging this one.

File: deploy/role.yaml
Comment (laurafitzgerald) [inline deploy/role.yaml]: This is a copy of what's in https://github.com/aerogear/app-metrics-operator/pull/2

File: deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml
Comment (aliok) [inline deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml]: The kind looks a bit strange :)

Was it a mistake? Or if on purpose, can you explain what should resource do please?

Comment (aliok) [reply deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml]: Ah, this would represent a client app...
Maybe AppMetricsClient is a better name?
Or AppMetricsSource?

Comment (laurafitzgerald) [reply deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml]: I'm basing it on the MSS consumption cr kind which is MobileSecurityServiceApp, aka ServiceNameApp. It just happens that the name of this service has App in it already. Not sure I like it either. Thoughts on an alternative? In ups we use `variant` because that's what's created in the service instead of `app` which is what's created in the MSS service. There's no service here where something is created because of the bind. All we need to do is create the config map for use by MDC in mobile-services.json.

Comment (grdryn) [reply deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml]: Yeah, I'm not sure about the name of the installation CR either (AppMetricsService), happy to change that to if it helps.

@laurafitzgerald also note that in #2 I've changed the apiVersion to use .org instead of .com, to match with the other operators (I accidentally had it as .com when I created the project).

**Comment by aliok** [comment]
@laurafitzgerald Not related to your PR, but I noticed these env var names are stolen badly :)

https://github.com/aerogear/app-metrics-operator/blob/55158c28ad2380e16e94ba1a1aca29ea05983cff/pkg/config/config.go#L27-L28

Can you also fix that?

**Comment by laurafitzgerald** [comment]
@aliok sure, I'll fix them in a seperate pr.

**Comment by grdryn** [comment]
Oops, sorry about that :see_no_evil:

**Comment by laurafitzgerald** [comment]
@grdryn @aliok this is now ready for final review and verification. The steps are updated above.

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [inline pkg/controller/appmetricsapp/appmetricsapp_controller.go]: The two functions here are lifted from MSS.

File: Gopkg.lock
Comment (grdryn) [inline Gopkg.lock]: this looks like a merge/rebase conflict marker

File: Makefile
Comment (grdryn) [inline Makefile]: Could this be generated instead of hardcoded? If it's hardcoded we have to change it every time. I tend to use `git rev-parse --short HEAD`

File: deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml
Comment (grdryn) [reply deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml]: How about `AppMetricsConfig`? :thinking:

File: deploy/operator.yaml
Comment (grdryn) [reply deploy/operator.yaml]: If you don't want to have to do this for PRs, you can just put this in the verification steps:

```
sed -e "s|quay.io/aerogear/app-metrics-operator:master|quay.io/aerogear/app-metrics-operator:ag-9184|g" deploy/operator.yaml | kubectl create -f -
```

Or if you wanted a make task, something like this could go in there:
```
sed -e "s|quay.io/aerogear/app-metrics-operator:master|quay.io/aerogear/app-metrics-operator:$(git rev-parse --short HEAD)|g" deploy/operator.yaml | kubectl create -f -
```

File: pkg/apis/metrics/v1alpha1/appmetricsapp_types.go
Comment (grdryn) [inline pkg/apis/metrics/v1alpha1/appmetricsapp_types.go]: Could you remove these boiler-plate comments? If you add a custom comment, it should get added to the CRD I think.

File: pkg/apis/metrics/v1alpha1/appmetricsapp_types.go
Comment (grdryn) [inline pkg/apis/metrics/v1alpha1/appmetricsapp_types.go]: Is there an advantage to having a `spec["app-name"]` field, when there's the `metadata.name` field that must be specified anyway? If someone changes this, the name of the CR can't be changed, so they'd be out of sync.

If it's necessary, I'd suggest changing the json tag to be `appName`, since that's more normal/consistent in resource definitions (it can also be icky in some programming languages to reference fields by name if the name has a hyphen in it E.g. `spec["app-name"]` is needed instead of `spec.appName` in e.g. JavaScript, and it's not outside the realm of possibility that the MDC would read this in some future version)

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (grdryn) [inline pkg/controller/appmetricsapp/appmetricsapp_controller.go]: this todo comment can be removed

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (grdryn) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: I'd change them both to an initial lowercase letter, since they don't need to be exported from this package.

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (grdryn) [inline pkg/controller/appmetricsapp/appmetricsapp_controller.go]: This comment could do with some love: it has the function name wrong, and I don't know what it means by 'informed'

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (grdryn) [inline pkg/controller/appmetricsapp/appmetricsapp_controller.go]: The fact that this string gets split on semi-colon rather than comma or space or comma-space, was not obvious to me until now. :crying_cat_face: Maybe there should be a yaml comment in the `operator.yaml` (and README)? I'd suggest changing it to comma, but that probably makes it inconsistent with the other operator you took it from, right?

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (grdryn) [inline pkg/controller/appmetricsapp/appmetricsapp_controller.go]: the `IsValidAppNamespace` only ever seems to return `true, nil` or `false, err`, which makes the first boolean return value (which we clearly had trouble naming :sweat_smile:) and this check redundant I think (in the current implementation of that function, this block will never be entered if execution gets this far).

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (grdryn) [inline pkg/controller/appmetricsapp/appmetricsapp_controller.go]: I'd put this empty line above the previous line (assign to `err` and check its value in the same visual block). 

Actually I think that line 115 above could even be part of the `if` construct on line 117, since the only thing returned there is the `err` that will be checked and not used outside that check.

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (grdryn) [inline pkg/controller/appmetricsapp/appmetricsapp_controller.go]: I think this should be `cr.Namespace` rather than a hard-coded assumption (we only want to put the config map into a namespace called "mobile-developer-console" if that's where a CR was created, right?).

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (grdryn) [inline pkg/controller/appmetricsapp/appmetricsapp_controller.go]: This is an oddly specific version of a more generic function that could have the signature `func getEnvVar(envVar string) (string, error)`, which may or may not exist. My guess is that it doesn't exist because it's not really necessary.

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (grdryn) [inline pkg/controller/appmetricsapp/appmetricsapp_controller.go]: These two lines could be replaced with the following (removing the need for the `GetAppNamespaces()` function at all):

```go
appNamespacesEnvVar, found := os.LookupEnv("APP_NAMESPACES")
if !found {
```

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (grdryn) [inline pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Rather than hard-coding `"app-metrics"` here, I'd use `k8sutil.GetOperatorNamespace()` (not sure what that function does for `operator-sdk up local` though).

File: Makefile
Comment (grdryn) [inline Makefile:2]: this is different to the default in the operator.yaml, not sure if that's important.

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (grdryn) [inline pkg/controller/appmetricsapp/appmetricsapp_controller.go]: This is fine for now, but if we add more than this single field to the JSON object, I'd consider using something from the json package rather than hand-crafting it and escaping all of those quotes.

**Review by grdryn — DISMISSED** [review]
In case it's hidden (because that file doesn't show by default), there are conflict markers in Gopkg.lock in this PR that need to be removed. Further comments inline, but generally good and in the right direction! :+1:

File: Makefile
Comment (laurafitzgerald) [reply Makefile:2]: You are correct. :) It's done this way so that the operator can be decoupled from MDC in the community if someone wishes to do that. I think I need to update the value passed to the operator env var, though 🤔

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: 👍 copy pasta. Will improve.

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Yes, I used the semi-colon as it was used elsewhere. I prefer a comma also myself. I'm also unsure about the use of a list as our use case is for one namespace. Perhaps there are users upstream who would prefer to be able to provide a list but it hasn't been explicitly asked for yet. Interested in your thoughts also there.

File: Makefile
Comment (laurafitzgerald) [reply Makefile]: Yes, that shouldn't have been committed. the dev tag can be `dev`.

File: pkg/apis/metrics/v1alpha1/appmetricsapp_types.go
Comment (laurafitzgerald) [reply pkg/apis/metrics/v1alpha1/appmetricsapp_types.go]: Yes, we had a separate appName defined in the spec, the reason for this was that it could be updated in the MDC, which (I believe) metadata.name is not updateable. There was a discussion about removing the ability to update the app name in the MDC which I believe we are doing now so that may remove the need for this additional spec. But we should check that with @ziccardi or @craicoverflow before removing it. For now I will update to the better syntax until we can get some guidance from the guys.

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Ok. I'll recheck this. It's a copy pasta so I just took it as is. Happy to make improvements. :)

File: deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml
Comment (aliok) [reply deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml]: sounds better than AppMetricsApp :) +1 on that

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: yes.

File: pkg/apis/metrics/v1alpha1/appmetricsapp_types.go
Comment (ziccardi) [reply pkg/apis/metrics/v1alpha1/appmetricsapp_types.go]: @grdryn @laurafitzgerald Correct. We already removed the ability to update the app name in the current master.

**Comment by laurafitzgerald** [comment]
@grdryn the two most recent commits are based on your feedback items. Could you take a look when you have some time?

**Comment by aliok** [comment]
Gave this PR a try locally.

I think it all works correctly, but there's some incompatibility with what's written in the verification instructions. Maybe they're outdated?

File: deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml
Comment (laurafitzgerald) [reply deploy/crds/metrics_v1alpha1_appmetricsapp_cr.yaml]: +1 Going with `AppMetricsConfig`, will push the change soon.

File: pkg/apis/metrics/v1alpha1/appmetricsapp_types.go
Comment (laurafitzgerald) [reply pkg/apis/metrics/v1alpha1/appmetricsapp_types.go]: So we should not use spec.appName?

Comment (laurafitzgerald) [reply pkg/apis/metrics/v1alpha1/appmetricsapp_types.go]: @ziccardi

File: deploy/operator.yaml
Comment (grdryn) [inline deploy/operator.yaml]:
```suggestion
              # Replace this with a ; separated list of namespaces in which the operator will act on app crs
```

File: pkg/apis/metrics/v1alpha1/appmetricsapp_types.go
Comment (grdryn) [inline pkg/apis/metrics/v1alpha1/appmetricsapp_types.go]: It sounds like based on the conversation in the previous review, that this can be removed. Wdyt?

Comment (laurafitzgerald) [reply pkg/apis/metrics/v1alpha1/appmetricsapp_types.go]: Removed. 👍

File: deploy/operator.yaml
Comment (laurafitzgerald) [reply deploy/operator.yaml]: Done

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Done

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Done

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Done

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Done

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Done

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Done

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Changed to comma. Done

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Done

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Done

File: pkg/apis/metrics/v1alpha1/appmetricsapp_types.go
Comment (laurafitzgerald) [reply pkg/apis/metrics/v1alpha1/appmetricsapp_types.go]: app-name removed. Done

File: pkg/apis/metrics/v1alpha1/appmetricsapp_types.go
Comment (laurafitzgerald) [reply pkg/apis/metrics/v1alpha1/appmetricsapp_types.go]: Done

File: Makefile
Comment (laurafitzgerald) [reply Makefile]: Done

File: Gopkg.lock
Comment (laurafitzgerald) [reply Gopkg.lock]: Done

File: pkg/controller/appmetricsapp/appmetricsapp_controller.go
Comment (laurafitzgerald) [reply pkg/controller/appmetricsapp/appmetricsapp_controller.go]: Done

**Comment by laurafitzgerald** [comment]
@aliok yes those were out of date. They are up to date now and I've pushed additional changes as suggested.  All the current feedback is addressed so this is ready to be verified now. Thanks for all the feedback. :)

**Review by grdryn — APPROVED** [review]
All looks good to me: reviewed and verified.

One thing from the verification steps is that the log message was different than in the steps, but I don't think that's a big deal:

```
{"level":"error","ts":1562689049.299172,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"appmetricsconfig-controller","request":"app-metrics/example-app","error":"The app cr example-app was created in a namespace which is not present in the APP_NAMESPACES environment variable provided to the operator","stacktrace":"github.com/aerogear/app-metrics-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
```

Additionally, in the code, the `DEV_TAG` stuff still exists and/or is specific to this PR.

**Comment by laurafitzgerald** [comment]
> All looks good to me: reviewed and verified.
> 
> One thing from the verification steps is that the log message was different than in the steps, but I don't think that's a big deal:
> 
> ```
> {"level":"error","ts":1562689049.299172,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"appmetricsconfig-controller","request":"app-metrics/example-app","error":"The app cr example-app was created in a namespace which is not present in the APP_NAMESPACES environment variable provided to the operator","stacktrace":"github.com/aerogear/app-metrics-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/laurafitzgerald/go/src/github.com/aerogear/app-metrics-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
> ```
> 
> Additionally, in the code, the `DEV_TAG` stuff still exists and/or is specific to this PR.

@grdryn thanks for the approval. I will remove the DEV_TAG stuff and I've updated the description for the verification.  I changed the log message when I made the improvements.

**Comment by laurafitzgerald** [comment]
I've cleaned up the commits and reverified everything works my end. Merging as it has approval.

**Pull request merged**
